docs/dev/stream/operators/joining.md

```diff
@@ -0,0 +1,286 @@
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+
+```java
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.streaming.api.windowing.assigners.EventTimeSessionWindows;
+import org.apache.flink.streaming.api.windowing.time.Time;
+ 
+...
+
+DataStream<Integer> orangeStream = ...
+DataStream<Integer> greenStream = ...
+
+orangeStream.join(greenStream)
+    .where(<KeySelector>)
+    .equalTo(<KeySelector>)
+    .window(EventTimeSessionWindows.withGap(Time.milliseconds(1)))
+    .apply (new JoinFunction<Integer, Integer, String> () {
+        @Override
+        public String join(Integer first, Integer second) {
+            return first + "," + second;
+        }
+    });
+```
+</div>
+<div data-lang="scala" markdown="1">
+
+```scala
+import org.apache.flink.streaming.api.windowing.assigners.EventTimeSessionWindows;
+import org.apache.flink.streaming.api.windowing.time.Time;
+ 
+...
+
+val orangeStream: DataStream[Integer] = ...
+val greenStream: DataStream[Integer] = ...
+
+orangeStream.join(greenStream)
+    .where(elem => /* select key */)
+    .equalTo(elem => /* select key */)
+    .window(EventTimeSessionWindows.withGap(Time.milliseconds(1)))
+    .apply { (e1, e2) => e1 + "," + e2 }
+```
+
+</div>
+</div>
+
+# Interval Join
+The interval join joins elements of two streams (we'll call them A & B for now) with a common key and where elements of stream B have timestamps that lie in a relative time interval to timestamps of elements in stream A.
+
+This can also expressed a little more formally as 
+`b.timestamp ∈ [a.timestamp + lowerBound; a.timestamp + upperBound]` or 
+`a.timestamp + lowerBound <= b.timestamp <= a.timestamp + upperBound`
+
+where a and b are elements of A and B that share a common key.
+
+The interval join currently only performs inner joins, meaning that if an element from one stream has no element from the other stream to be joined with, it will not be passed to the user-defined function. 
+
+When joined elements are passed to the user-defined function they have the maximum timestamp of either one of the two elements. The timestamps of either one of those elements can also be accessed via the `ProcessJoinFunction.Context`
+
+Note: The interval join currently only supports event time.
+
+<img src="{{ site.baseurl }}/fig/interval-join.svg" class="center" style="width: 80%;" />
+
+In this example we are joining two streams 'orange' and 'green' with a lower bound of minus two milliseconds and an upper bound of one millisecond. By default those boundaries are inclusive, but `.lowerBoundExclusive()` and / or `.upperBoundExclusive()` can be used to specify different behaviour.
```

Comment:
In the example above, we join two streams 'orange' and 'green' with a lower bound of -2 milliseconds and an upper bound of +1 millisecond. By default, those boundaries are inclusive, but `.lowerBoundExclusive()` and `.upperBoundExclusive()` can be applied to change the behavior.